scanpy/get/aggregated.py

```diff
@@ -0,0 +1,401 @@
+        arr[:] = lst
+    else:
+        arr = np.array(lst)
+    return arr
+
+
+def _superset_columns(df: pd.DataFrame, groupby_key: str) -> List[str]:
+    """Find all columns which are a superset of the key column.
+
+    Args:
+        df (pd.DataFrame): DataFrame which contains candidate columns.
+        groupby_key (str): Key for column of which to find superset of columns.
+
+    Returns:
+        List[str]: Superset columns.
+    """
+    columns = []
+    groupy_key_codes = df[groupby_key].astype('category')
+    for key in df:
+        if key != groupby_key:
+            key_codes = df[key].astype('category')
+            if all(
+                [
+                    key_codes[groupy_key_codes == group_key_code].nunique() == 1
+                    for group_key_code in groupy_key_codes
+                ]
+            ):
+                columns += [key]
+    return columns
+
+
+def _df_grouped(df: pd.DataFrame, key: str, key_set: List[str]) -> pd.DataFrame:
+    """Generate a grouped-by dataframe (no aggregation) by a key with columns that are supersets of the key column
+
+    Args:
+        df (pd.DataFrame): DataFrame to be grouped.
+        key (str): Column to be grouped on.
+        key_set (List[str]): values in the `key` column to keep before groupby.
+
+    Returns:
+        pd.DataFrame: Grouped-by Dataframe.
+    """
+    df = df.copy()
+    if key_set is not None:
+        df = df[df[key].isin(key_set)]
+    if df[key].dtype.name == 'category':
+        df[key] = df[key].cat.remove_unused_categories()
+    return df.groupby(key).first()[_superset_columns(df, key)]
+
+
+@singledispatch
+def aggregated(
+    adata: AnnData,
+    by: str,
+    how: Literal['count', 'mean', 'sum', 'count_mean_var'] = 'count_mean_var',
+    groupby_df_key: Literal['obs', 'var'] = 'obs',
+    weight_key: Optional[str] = None,
+    key_set: Optional[Iterable[str]] = None,
+    dof: int = 1,
+    layer: Optional[str] = None,
+    obsm: Optional[str] = None,
+    varm: Optional[str] = None,
+) -> AnnData:
+    """Aggregate data based on one of the columns of one of the axes (`obs` or `var`).  If none of `layer`, `obsm`, or `varm` are passed in, `X` will be used for aggregation data.
+
+    Args:
+        adata (AnnData): AnnData to be aggregated.
+        by (str): Key of the column to be grouped-by.
+        how (Literal[&#39;count&#39;, &#39;mean&#39;, &#39;sum&#39;, &#39;count_mean_var&#39;], optional):  How to aggregate. Defaults to 'count_mean_var'.
+        groupby_df_key (Literal[&#39;obs&#39;, &#39;var&#39;], optional): Axis on which to find group by column. Defaults to 'obs'.
+        weight_key (Optional[str], optional): Key of the `groupby_df_key` containing weights for a weighted sum aggregation. Defaults to None.
+        key_set (Optional[Iterable[str]], optional): Subset of groupby_df_key on which to filter. Defaults to None.
+        dof (int, optional): Degrees of freedom for variance. Defaults to 1.
```

Comment:
I'm wondering if we can remove this. I think this would only ever apply to `var`, and the value should almost always be 1, maybe sometimes 0. I would even consider just say `var` and `var_ddof0`. I think I've seen other packages do this.

Reply:
Alright I can do that.